src/main/java/com/werfen/report/model/GridReportDataSource.java

```diff
@@ -11,35 +10,40 @@
 
 public class GridReportDataSource extends JRAbstractBeanDataSource {
 
-    private final GridReportData rows;
+    private final GridPageDataSource dataSource;
     private Iterator<GridReportRow> rowIterator;
     private GridReportRow currentRow;
 
-    public GridReportDataSource(GridReportData rows) {
+    public GridReportDataSource(GridPageDataSource dataSource) {
         super(true);
-        this.rows = rows;
-        if (nonNull(rows)) {
-            this.rowIterator = this.rows.getRows().listIterator();
+        this.dataSource = dataSource;
+        if (this.dataSource.getRowCount() > 0) {
+            this.dataSource.moveFirst();
         }
     }
 
+
     @Override
     public void moveFirst() {
-        if (nonNull(rows)) {
-            rowIterator = rows.getRows().listIterator();
+        if (dataSource.getRowCount() > 0) {
```

Comment:
We don't need to check if we have at least one row using row count. We can do so by checking the list returned by the data source for the current page:

```java
this.dataSource.moveFirst();
List<GridReportRow> currentPageRows =  this.dataSource.getCurrentPageRows();
if (currentPageRows.size() > 0) {
    this.currentPageRowIterator = currentPageRows.listIterator();
    this.currentRow = this.currentPageRowIterator .next();
}
```

Additionally, as in the code snippet, consider renaming _rowIterator_ as _currentPageRowIterator_, so we have a better understanding of what iterator refers to.